### PR TITLE
BSH tweak/refactor/fix

### PR DESCRIPTION
**Pull request by Migratingcocofruit**
## What Does This PR Do
Removes the level abstraction from the BSH, instead taking direct power input, which is unlimited, refactors the process logic to be a bit more streamlined, fixes a couple oversights to do with portal spawning not being increased by high power as much as it should have been and adds a little UI overlay when the BSH encounters an incursion.
Point production and incursion chance remains the same for the same power level, however the amount of portals does increase at higher levels as a result of fixing an oversight from #26032

## Why It's Good For The Game
The BSH is the powersink goal, there isn't anything that can consume more power, and yet as things stand a single SM can produce enough to fully supply it, even when stabilized. 
In addition after moving to mining points per MW in #26032 it only makes sense to do away with the level abstraction as it doesn't have much use anymore. 

## Images of changes
image
ezgif-2-0d2cd3b0f6

## Testing
-Verified all the inputs do what you expect them to
-Verified the overlay comes on and off when it should
-Verified the correct amount of portals is being spawned when it should be spawned
-Ran without stabilizers or auto shut down at 2GW and drowned in a sea of mobs

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

:cl:
add: UI overlay that triggers when there is an incursion
tweak: Replaced level system with direct power input. 
fix: Fixed a bug that caused an insufficient amount of additional portals spawn at higher power levels
/:cl:

### CODE REVIEW

File: code/modules/station_goals/bluespace_tap.dm [hunk 1]
```diff
@@ -296,18 +285,18 @@
 			underlays += emissive_appearance(icon, "light_mask")
 
 /obj/machinery/power/bluespace_tap/proc/get_icon_state_number()
-	switch(input_level)
+	switch(mining_power)
 		if(0)
 			return 0
-		if(1 to 3)
+		if(50 KW to 3 MW)
```
Comment (lewcc) [inline code/modules/station_goals/bluespace_tap.dm:291]: I think you'd want to extend this range for 0, or just make it the else condition.

File: code/modules/station_goals/bluespace_tap.dm [hunk 2]
```diff
@@ -296,18 +285,18 @@
...
 			return 1
-		if(4 to 8)
+		if(4 MW to 8 MW)
 			return 2
-		if(9 to 11)
+		if(9 MW to 11 MW)
```
Comment (lewcc) [inline code/modules/station_goals/bluespace_tap.dm:295]: I'm pretty sure that if you have 8.1 MW, you'll fall out of this. Make sure the full range is captured.

File: code/modules/station_goals/bluespace_tap.dm [hunk 3]
```diff
@@ -331,120 +320,97 @@
...
+	// Stabilizers consume up to 1MW for each 1MW of mining power, consuming less between 15 and 30MW of mining power
+	// If stabilizers have priority they will always consume enough power to stabilize the BSH, limiting mining
+	// emagging disables stabilizers
+	if(stabilizer_priority)
+		stabilizer_power = clamp(desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW), 0, mining_power / 2) * (stabilizers && !emagged)
+	else
+		stabilizer_power = clamp(mining_power - desired_mining_power, 0, desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW)) * (stabilizers && !emagged)
+
+	// Actual mining power is what the desired mining power we set, unless we don't have enough power to satisfty that.
+	mining_power = min(desired_mining_power , mining_power - stabilizer_power)
```
Comment (lewcc) [inline code/modules/station_goals/bluespace_tap.dm:361]:
```suggestion
	mining_power = min(desired_mining_power, mining_power - stabilizer_power)
```

File: code/modules/station_goals/bluespace_tap.dm [hunk 4]
```diff
@@ -331,120 +320,97 @@
...
+	// Always try to use all available power for mining if emagged
+	if(emagged)
+		desired_mining_power = mining_power
+
+	// Stabilizers activate above 15MW of mining power
+	// Stabilizers consume up to 1MW for each 1MW of mining power, consuming less between 15 and 30MW of mining power
+	// If stabilizers have priority they will always consume enough power to stabilize the BSH, limiting mining
+	// emagging disables stabilizers
+	if(stabilizer_priority)
+		stabilizer_power = clamp(desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW), 0, mining_power / 2) * (stabilizers && !emagged)
```
Comment (lewcc) [inline code/modules/station_goals/bluespace_tap.dm:356]: Do you think you could break these calculations up into a few lines so it's a little clearer what's going on?

File: code/modules/station_goals/bluespace_tap.dm [hunk 5]
```diff
@@ -331,120 +320,97 @@
...
+	else
+		stabilizer_power = clamp(mining_power - desired_mining_power, 0, desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW)) * (stabilizers && !emagged)
+
+	// Actual mining power is what the desired mining power we set, unless we don't have enough power to satisfty that.
+	mining_power = min(desired_mining_power , mining_power - stabilizer_power)
+
+	consume_direct_power(mining_power + stabilizer_power)
+
+	// 2 points per 50 KW up to 20 and 4 points per MW (or 5 when emmaged).
+	mined_points = min(BASE_POINTS * (mining_power / (50 KW)) , 20) + mining_power * (POINTS_PER_W + emagged / (1 MW))
```
Comment (lewcc) [inline code/modules/station_goals/bluespace_tap.dm:366]:
```suggestion
	mined_points = min(BASE_POINTS * (mining_power / (50 KW)), 20) + mining_power * (POINTS_PER_W + emagged / (1 MW))
```

Also, did you mean to have parentheses in the last term? Just checking

Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:366]: Yes.

File: code/modules/station_goals/bluespace_tap.dm (hunk 4 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:356]: I'll try. I'll explain it in a comment anyhow

Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:356]: Byond complains when I break it into different lines. I added a comment for each calculation, and there is additional clarification above

File: code/modules/station_goals/bluespace_tap.dm (hunk 2 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:295]: When above 1MW mining power is rounded down to the nearest MW

Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:295]: I suppose it wouldn't matter if I did capture the full range anyhow

File: code/modules/station_goals/bluespace_tap.dm (hunk 4 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:357]: Done

File: code/modules/station_goals/bluespace_tap.dm [hunk 6]
```diff
@@ -331,120 +320,108 @@
...
+		mining_power = mining_power - mining_power % (1 MW)
+
+	// Always try to use all available power for mining if emagged
+	if(emagged)
+		desired_mining_power = mining_power
+
+	// Stabilizers activate above 15MW of mining power
+	// Stabilizers consume up to 1MW for each 1MW of mining power, consuming less between 15 and 30MW of mining power
+	// If stabilizers have priority they will always consume enough power to stabilize the BSH, limiting mining
+	// emagging disables stabilizers
```
Comment (Burzah) [inline code/modules/station_goals/bluespace_tap.dm:354]:
```suggestion
	/* 
	* Stabilizers activate above 15MW of mining power
	* Stabilizers consume up to 1MW for each 1MW of mining power, consuming less between 15 and 30MW of mining power
	* If stabilizers have priority they will always consume enough power to stabilize the BSH, limiting mining
	* Emagging disables stabilizers
	*/
```
Multiline comment format.

File: code/modules/station_goals/bluespace_tap.dm [hunk 7]
```diff
@@ -331,120 +320,108 @@
...
+	// 2 points per 50 KW up to 20 and 4 points per MW (or 5 when emmaged).
+	mined_points = min(BASE_POINTS * (mining_power / (50 KW)) , 20) + mining_power * (POINTS_PER_W + emagged / (1 MW))
+	points += mined_points
+	total_points += mined_points
+	update_icon()
+	// Portal chance is 0.1% per cycle per difference of 1MW between the stabilizer and mining power
+	// This should translate to portals spawning every 33:20 minutes for 1 difference of 1MW
+	// And about every 3:20 minutes for a 10MW difference
+	// Emagging garuantees a chance of at least 5%.
+	// Prob treats values less than 0 as 0.
```
Comment (Burzah) [inline code/modules/station_goals/bluespace_tap.dm:385]:
```suggestion
	/*
	* Portal chance is 0.1% per cycle per difference of 1MW between the stabilizer and mining power
	* This should translate to portals spawning every 33:20 minutes for 1 difference of 1MW
	* And about every 3:20 minutes for a 10MW difference
	* Emagging guarantees a chance of at least 5%
	* Prob treats values less than 0 as 0.
	*/
```

File: code/modules/station_goals/bluespace_tap.dm [hunk 8]
```diff
@@ -174,7 +174,9 @@
 		/obj/item/food/sliceable/xenomeatbread //maybe add some dangerous/special food here, ie robobuger?
 	)
 
-#define BASE_ENERGY_CONVERSION 4e-6
+///Points generated per cycle for each Watt of power consumption
+#define POINTS_PER_W 4e-6
+/// amount of points generated per cycle per 50KW for the first 500KW
```
Comment (Burzah) [inline code/modules/station_goals/bluespace_tap.dm:179]:
```suggestion
/// Points generated per cycle for each Watt of power consumption
#define POINTS_PER_W 4e-6
/// Amount of points generated per cycle per 50KW for the first 500KW
```

File: code/modules/station_goals/bluespace_tap.dm (hunk 7 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm]: done

File: code/modules/station_goals/bluespace_tap.dm (hunk 6 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm]: done

File: code/modules/station_goals/bluespace_tap.dm (hunk 8 above)
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm]: done

File: code/modules/station_goals/bluespace_tap.dm [hunk 9]
```diff
@@ -331,120 +320,114 @@
...
+		0, \
+		mining_power / 2) \
+		* (stabilizers && !emagged)
+	else
+		// stabilizer power is however much power we have left, but no more than we need to stabilize our desired mining power.
+		stabilizer_power = \
+		clamp(mining_power - desired_mining_power, \
+		0, \
+		desired_mining_power - clamp(30 MW - desired_mining_power, 0, 15 MW)) \
+		* (stabilizers && !emagged)
```
[the remark this answers is not in the document]
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm:371]: image

File: tgui/packages/tgui/interfaces/BluespaceTap.js [hunk 10]
```diff
@@ -1,30 +1,68 @@
...
+          interval={Math.random() > 0.25 ? 750 + 400 * Math.random() : 290 + 150 * Math.random()}
+          time={60 + 150 * Math.random()}
+        >
+          <Flex mb="30px" fontsize="256px">
+            <Flex.Item bold color="red" fontsize="256px" textAlign="center">
+              <Icon name="skull" size={14} mb="64px" />
+              <br />
+              E$#OR:& U#KN!WN IN%ERF#R_NCE
+            </Flex.Item>
+          </Flex>
```
[the remark this answers is not in the document]
Comment (Migratingcocofruit) [reply tgui/packages/tgui/interfaces/BluespaceTap.js:37]: I'll give it a shot

[the remark this answers is not in the document]
Comment (Migratingcocofruit) [reply tgui/packages/tgui/interfaces/BluespaceTap.js]: done

File: code/modules/station_goals/bluespace_tap.dm [hunk 11]
```diff
@@ -331,120 +320,114 @@
...
+	update_icon()
+	/*
+	* Portal chance is 0.1% per cycle per difference of 1MW between the stabilizer and mining power
+	* This should translate to portals spawning every 33:20 minutes for 1 difference of 1MW
+	* And about every 3:20 minutes for a 10MW difference
+	* Emagging guarantees a chance of at least 5%
+	* Prob treats values less than 0 as 0.
+	*/
+	// Emagging garuantees a chance of at least 5%.
+	// Prob treats values less than 0 as 0.
```
[the remark this answers is not in the document]
Comment (Migratingcocofruit) [reply code/modules/station_goals/bluespace_tap.dm]: done

**Pull request merged**